### ISSUE TITLE
Windows Server user was given access to the RStudio Desktop session of another user

### ISSUE THREAD

**Issue by WetRobot** [rstudio/rstudio#10488]
### System details

    RStudio Edition : Desktop
    RStudio Version : RStudio 2021.09.0+351 "Ghost Orchid" Release (077589bcad3467ae79f318afe8641a1899a51606, 2021-09-20) for Windows Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36
    OS Version      : Microsoft Windows Server 2012 R2 Standard
    R Version       : 4.1.1 (2021-08-10) -- "Kick Things"

### Steps to reproduce the problem

I am unable to give steps to reproduce this because of its random nature. In many years of using RStudio Desktop on a Windows Server with the projects on a network drive, the issue has only occurred once. The issue may potentially be reproduced by forcing two RStudio Desktop sessions to use the same port, which I at least was unable to do.

#### What we know

Users A and B logged onto Windows Server 2012. User A had a running RStudio Desktop session from the previous day. User B is uncertain whether they had a previous session open or whether they opened a new RStudio Desktop session upon log-in. User A had their project on a network drive. User A saw upon log-in that Rstudio Desktop session had complained of failed file monitoring (related: https://github.com/rstudio/rstudio/issues/2414):

```
Warning message:
File monitoring failed for project at "X:/path/to/network/drive/location"
Error 2 (The system cannot find the file specified)
Features disabled: R source file indexing, Diagnostics
```

RStudio Desktop worked normally for user A. However, the RStudio Desktop session for user B was that of user A: User B was able to see the session of user A and even to run code in it. Our IT has confirmed that Windows processes themselves were not shared. We have not encountered an issue like this with any other software.

#### Speculation 

Because --- to my understanding --- RStudio Desktop is essentially a web browser and shows a “web page” in some port of 127.0.0.1, the problem may have come from RStudio Desktop opening the same “web page” for user B which user A created for their session. To my (limited) understanding this can theoretically happen if both users end up with the same port, which RStudio appears to select randomly (https://github.com/rstudio/rstudio/blob/edd144e6c64f765f62716e57a31808597e92ab7e/src/node/desktop/src/core/system.ts#L34). I speculate that user B had opened a new session of RStudio Desktop, which selected the same port which user A had in their session. The file monitoring issue is in my view unrelated, but I have included that information here because that is what user A saw in their session when they logged in.

### Describe the behavior you expected

Two users logged into the same computer (e.g. Windows Server) should never be able to see and interact with the RStudio Desktop sessions of other users without explicit permission. The same session being shared to another user without their knowledge is a substantial security problem for us, because we regularly process sensitive information in RStudio.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by mikebessuille** [comment]
Probably relatively few users encounter this, but the results are serious.  Need to investigate in spotted wakerobin

**Comment by jeffvroom** [comment]
The code link above is part of the unreleased electron version, but looks like we do something similar [here]( https://github.com/rstudio/rstudio/blob/26c12963b13f05512cffcb3805bb727c9c6d05c2/src/cpp/desktop/DesktopOptions.cpp#L159) for desktop.  It looks like there's also a sharedSecret that's passed from the desktop client to the session it starts that should prevent one user's sessions from being able to attach to another's. I agree we should make sure this is not possible!

**Comment by kevinushey** [comment]
That code hasn't changed in a long time, though -- if that code is the culprit, then it's quite surprising to me that we're only seeing it now and not with previous releases of RStudio. That said, I'm not sure where else we would want to look to try and reproduce.

**Comment by kevinushey** [comment]
That said ... I was able to reproduce locally, just by launching a bunch of RStudio instances in quick succession on Windows, so this issue is definitely real!

**Comment by kevinushey** [comment]
If I try the same test on macOS, I see a bunch of RStudio instances failing to launch with an error like:

> 2022-04-01T18:20:06.681591Z [rsession-kevin] ERROR system error 48 (Address already in use); OCCURRED AT rstudio::core::Error rstudio::core::http::initTcpIpAcceptor(SocketAcceptorService<boost::asio::ip::tcp> &, const std::string &, const std::string &) src/cpp/session/http/SessionPosixHttpConnectionListener.cpp:103; LOGGED FROM: int main(int, char *const *) src/cpp/session/SessionMain.cpp:2142

It seems a bit like the different IDE instances are racing to communicate with and take "ownership" of the newly-launched `rsession` executable. On macOS, only once instance succeeds; on Windows, it seems like multiple front-ends are able to successfully bind to a single `rsession` instance.

**Comment by kevinushey** [comment]
Indeed though, if I add some debug logging, it looks like all of these sessions end up trying to communicate on the same port + URL combination.

I wonder if a recent Boost update might've changed the behavior of the random number generator, or if we're using it incorrectly?

**Comment by kevinushey** [comment]
Note: still needs to be backported to PT, GO.

**Comment by mikebessuille** [comment]
@kevinushey please let Ron know when this has been backported.

**Comment by kevinushey** [comment]
Should now be backported to both PT + GO; builds are still in-flight as some compilation issues needed fixing.

**Comment by mikebessuille** [comment]
@kevinushey this issue still needs a NEWS entry for PT and GO patches I believe

**Comment by kevinushey** [comment]
Those should have associated NEWS entries; e.g.

PT NEWS: https://github.com/rstudio/rstudio/blob/50c401e8da4d4a5fbfcfaa0486d172fd6e9c1d47/version/news/NEWS-2022.02.2-prairie-trillium.md#L16

GO NEWS: https://github.com/rstudio/rstudio/blob/edce98a0ccd34abd2639658556f193a13c8231ee/NEWS-2021.09.3-ghost-orchid.md#L11

**Comment by ronblum** [comment]
Verified using the steps above on Windows 11 in RStudio Desktop
- Ghost Orchid: 2021.09.3-391
- Prairie Trillium: 2022.02-481
- Spotted Wakerobin 2022.06.0-dailiy-290

**Comment by atocharnaud** [comment]
> > We don't seems to experience this issue on more recent versions of Windows server.
> 
> To clarify, do you mean that newer versions of RStudio Desktop work fine with newer versions of Windows server?

Yes, I can not reproduce the issue I reported above using Windows 2012R2  on a Windows server 2019 VM.

Even when trying to run the command below, the port already in use error we now get with Rstudio 485 does not shows up on windows server 2019.
`for %i in (1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20) do start /b rstudio.exe`

I am able to reproduce the bug on windows server 2019 and windows server 2012R2 using Rstudio 443 starting only 3 instances.

**Comment by atocharnaud** [comment]
> Thanks for the bug report -- the main potential change that might be affecting you here is that RStudio now attempts to bind on a port in the range of 49152 to 65535, rather than the previous (which IIUC was something like 8080 to 48080). Is that port range open?

Ports above 49152 are open:
  TCP    127.0.0.1:60258        ECO-RDS-1:44554        ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60259        ECO-RDS-1:10888        ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60260        ECO-RDS-1:8789         ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60261        ECO-RDS-1:8308         ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60262        ECO-RDS-1:26838        ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60263        ECO-RDS-1:11710        ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:60264        ECO-RDS-1:41945        ESTABLISHED
 [rstudio.exe]
  TCP    127.0.0.1:62522        ECO-RDS-1:59012        ESTABLISHED

**Comment by wtimmerman-fitp** [comment]
For reference, myself and at least one other user encountered this Error 10048 and reported it on RStudio Community if any additional cases are handy: https://community.rstudio.com/t/error-10048-on-rstudio-startup/139278 . A

**Comment by mikebessuille** [comment]
@wtimmerman-fitp would you mind retesting on the latest patch version of Prairie Trillium or Ghost Orchid?  This has been fixed and release; See https://github.com/rstudio/rstudio/issues/10488#issuecomment-1096745272

**Comment by wtimmerman-fitp** [comment]
Sure thing, will check tomorrow!

**Comment by wtimmerman-fitp** [comment]
Interesting. These are the ones I've tried so far where the issue has appeared:

2021.09.3+391 (from daily builds, referenced in earlier reply)
2021.09.3-396 (official release)
2022.02.2+481 (from daily builds, referenced in earlier reply)
2022.02.3-492 (official release)
2022.06.0-daily-497 (from daily builds)

The electron version of Spotted Wakerobin didn't have the issue (2022.06.0-daily+496)

I did notice that on 2022.06.0-daily-497, i also received this error message in addition to the Error 10048:

`2022-06-09T12:30:02.868925Z [rsession-myname] ERROR system error 10053 (An established connection was aborted by the software in your host machine) [request-uri: /help/library/stringr/html/str_match.html]; OCCURRED AT void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\main\src\cpp\session\http\SessionHttpConnectionImpl.hpp:156; LOGGED FROM: void __cdecl rstudio::session::HttpConnectionImpl<class rstudio_boost::asio::ip::tcp>::sendResponse(const class rstudio::core::http::Response &) C:\Users\jenkins\workspace\ide\open-source-pipeline\main\src\cpp\session\http\SessionHttpConnectionImpl.hpp:161`

**Comment by wtimmerman-fitp** [comment]
Yes, I think a new issue is warranted (I can open that now) and indeed, i'm not having the original issue here. was just directed here from RStudio Community, thus the initial tag-on to this one.

**Comment by mikebessuille** [comment]
Great, thank you!  And thank you so much for restesting several versions to verify both the error and the original fix.  Much appreciated!
